src/nupic/research/temporal_memory.py

```diff
@@ -545,96 +407,77 @@ def leastUsedCell(self, cells, connections):
 
-    @param segment              (int)         Segment index
-    @param activeSynapses       (set)         Indices of active synapses
-    @param connections          (Connections) Connectivity of layer
+    @param prevActiveCells      (list)   Active cells in `t-1`
     @param permanenceIncrement  (float)  Amount to increment active synapses
     @param permanenceDecrement  (float)  Amount to decrement inactive synapses
+    @param segment              (int)    Segment to adapt
     """
     # Need to copy synapses for segment set below because it will be modified
     # during iteration by `destroySynapse`
-    for synapse in set(connections.synapsesForSegment(segment)):
-      synapseData = connections.dataForSynapse(synapse)
+    for synapse in set(self.connections.synapsesForSegment(segment)):
+      synapseData = self.connections.dataForSynapse(synapse)
       permanence = synapseData.permanence
 
-      if synapse in activeSynapses:
+      if self.binSearch(prevActiveCells, synapseData.presynapticCell) != -1:
         permanence += permanenceIncrement
       else:
         permanence -= permanenceDecrement
 
       # Keep permanence within min/max bounds
       permanence = max(0.0, min(1.0, permanence))
 
-      if (permanence < EPSILON):
-        connections.destroySynapse(synapse)
+      if permanence < EPSILON:
+        self.connections.destroySynapse(synapse)
       else:
-        connections.updateSynapsePermanence(synapse, permanence)
+        self.connections.updateSynapsePermanence(synapse, permanence)
 
+    # awaiting change to connections.py to facilitate deleting segments
+    # and synapses like the c++ implementation.
+    # if (len(self.connections.synapsesForSegment(segment)) == 0):
+    #   self.connections.destroySegment(segment)
 
-  def pickCellsToLearnOn(self, n, segment, winnerCells, connections):
-    """
-    Pick cells to form distal connections to.
-
-    TODO: Respect topology and learningRadius
-
-    @param n           (int)         Number of cells to pick
-    @param segment     (int)         Segment index
-    @param winnerCells (set)         Indices of winner cells in `t`
-    @param connections (Connections) Connectivity of layer
-
-    @return (set) Indices of cells picked
-    """
-    candidates = set(winnerCells)
-
-    # Remove cells that are already synapsed on by this segment
-    for synapse in connections.synapsesForSegment(segment):
-      synapseData = connections.dataForSynapse(synapse)
-      presynapticCell = synapseData.presynapticCell
-
-      if presynapticCell in candidates:
-        candidates.remove(presynapticCell)
-
-    n = min(n, len(candidates))
-    candidates = sorted(candidates)
-    cells = set()
-
-    # Pick n cells randomly
-    for _ in range(n):
-      i = self._random.getUInt32(len(candidates))
-      cells.add(candidates[i])
-      del candidates[i]
-
-    return cells
-
+  @staticmethod
+  def binSearch(arr, val):
```

Comment:
This seems a little bit out of place as a TemporalMemory method. Maybe just make it a function? I don't have a strong stance on this, but that's what I'd prefer.